### PR TITLE
*: improve 'insert into select' performance by optimize utf8 charset check

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Insert into select is very slow ...

```
CREATE TABLE `xx` (
  `id` int(11) DEFAULT NULL,
  `c` varchar(512) DEFAULT NULL
)
```
```
CREATE GLOBAL TEMPORARY TABLE `t` (
  `id` int(11) DEFAULT NULL,
  `c` varchar(512) DEFAULT NULL
)
```
prepare some data:

```
insert into xx values (1, repeat(‘x’, 512));
insert into xx select * from xx;  // execute several times
insert into xx select * from xx limit yy;
```

```
explain analyze insert into t select * from xx;
time:5.06s, prepare:3.1s, insert:1.95s  552.2 MB  1062144
```

As you can see, the total time is 5.06s, the actual insert cost **1.95s**, which prepare the data takes **3.1s** ...

image

A lot of the time is spend on the `CastValue` operation.

Issue Number: close #xxx 

Problem Summary:

### What is changed and how it works?

What's Changed:

In `CastValue`, we check the datum to be utf8mb4 encoding, it takes a lot of time.

I prefer to do that using the standard lib `utf8.ValidString(str)` for fast checking.
If it return false (which should be rare), we use the old way to check.

How it Works:

After this change, the time `insert into select` 550M down from **5s** to **3.32s**

And the flame graph now: 

image

### Check List 

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)

A benchmark test is added

```
cd session
 go test -bench BenchmarkInsertIntoSelect -run XXX -benchmem -cpuprofile cpu.out -benchtime 30s
```

Before vs After:

```
BenchmarkInsertIntoSelect-16    	     319	 112433885 ns/op	103542845 B/op	  365614 allocs/op
BenchmarkInsertIntoSelect-16    	     486	  73791669 ns/op	103508584 B/op	  365436 allocs/op
```

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

- No release note

### CODE REVIEW

File: table/column.go [hunk 1]
```diff
@@ -315,21 +315,23 @@ func CastValue(ctx sessionctx.Context, val types.Datum, col *model.ColumnInfo, r
 	str := casted.GetString()
 	utf8Charset := col.Charset == mysql.UTF8Charset
 	doMB4CharCheck := utf8Charset && config.GetGlobalConfig().CheckMb4ValueInUTF8
-	for i, w := 0, 0; i < len(str); i += w {
-		runeValue, width := utf8.DecodeRuneInString(str[i:])
-		if runeValue == utf8.RuneError {
-			if strings.HasPrefix(str[i:], string(utf8.RuneError)) {
-				w = width
-				continue
+	if !utf8.ValidString(str) {
```
Comment (tiancaiamao) [inline table/column.go:318]: This is the only line of changes...

**Review by xhebox — APPROVED** [review]
I checked [the source of validString](https://github.com/golang/go/blob/2ebe77a2fda1ee9ff6fd9a3e08933ad1ebaea039/src/unicode/utf8/utf8.go#L497), it looks identical, i.e. checking for 4 chars.

**Review by AilinKid — APPROVED** [review]
need cherry-pick ?

File: table/column.go (hunk 1 above)
Comment (wjhuang2016) [reply table/column.go:318]: If doMB4CharCheck is true and charset is utf8, a char with 4-bytes will insert successfully. But it shouldn't.

Comment (tiancaiamao) [reply table/column.go:318]: fixed

**Comment by tiancaiamao** [comment]
Close #25745

A regression test in the test case [show](http://www.zenlife.tk:18081/) the changes:
canvas

**Pull request merged**